### ISSUE TITLE
Defining segments based on arbitrary joins

### ISSUE THREAD

**Issue by bjeanes** [metabase/metabase#15564]
**Is your feature request related to a problem? Please describe.**
The majority of the questions we define for analytics or support issues continually re-define the same filters and conditions to view records in certain states ("approved", "rejected", etc, but also sometimes a few states will be bucketed together).

**Describe the solution you'd like**
When defining segments today, you can filter on associations it knows about at the schema level (i.e. foreign keys, whether or not they are _real_ FKs). In our case, there isn't always a FK for us to filter on. The fact that this ability works means that this segment _must_ be compiling to a join or subquery already.

The solution I have in mind is enriching the Segment UI to allow joining on arbitrary expressions and then applying filters to that joined data.

**Describe alternatives you've considered**
We are working around this by creating SQL questions which we can build other questions on, but this is not as discoverable as segments and fractures the places people must look. I am sure compelling alternatives may exist, but I don't have any in mind.

**How important is this feature to you?**
It's definitely only a nice-to-have, but seems like an obvious improvement for all users of segments.

**Additional context**
n/a

**Comment by flamber** [comment]
Hi @bjeanes 
So you're looking for the Notebook interface for Segments (and possible Metrics too) instead of the old QB UI?
I think an alternative would be to implement #5194, which would allow even more advanced structures.

**Comment by bjeanes** [comment]
Yes, though I am not sure everything in notebook UI is applicable. 

#5194 looks useful too, but we do have some people who would define segments but are not doesn't with SQL. But yes, it would be more flexible and perhaps a better starting point?

**Comment by thebiglabasky** [comment]
It looks like a duplicate of #17499
